README
///////////////
/// Welcome ///
///////////////
To get started, you'll need devIL (IL and ILU) and nVidia Texture Tools installed as shared libraries.
Then, hopefully, getting up and running is as simple as "make".

My apologies for the state of the code; this was never really intended to see the light of day, 
and it needs plenty of work.

~Cody Jung


/////////////////////////
/// Setting up prereqs //
/////////////////////////

First things, you'll need AT LEAST the following:
cmake, build-essential, and lib*-dev (e.g. libpng-dev) for each filetype you want to support

Next, download both Nvidia texture tools and the devIL libraries:
http://downloads.sourceforge.net/openil/DevIL-1.7.8.tar.gz
http://nvidia-texture-tools.googlecode.com/files/nvidia-texture-tools-2.0.8-1.tar.gz

	////////////////////////////
	/// Nvidia texture tools ///
	////////////////////////////
	
	After unzipping the tarball with "tar xvf nvidia-texture-tools-2.0.8-1.tar.gz", change to your new folder with cd nvidia-texture-tools.
	
	In this directory, type the following:
	$ ./configure
	$ make
	$ sudo make install
	
	/////////////
	/// DevIL ///
	/////////////
	
	Follow the same instructions as above for unzipping and getting to the right directory, then type:
	$ ./configure --enable-ILU --with-nvtt=no
	$ make
	$ sudo make install
	
If, while running VTFnix after compilation, you get the error:
	./VTFnix: error while loading shared libraries: libIL.so.1: cannot open shared object file: No such file or directory
	
First, make sure /usr/local/lib is listed in the file /etc/ld.so.conf (or one of it's includes), then run sudo /sbin/ldconfig

////////////////////
/// License Info ///
////////////////////
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

CHANGELOG
/////////////////
/// CHANGELOG ///
/////////////////

v. 1.0.1 - Bugfix
** Fixed animations not listing files in alpha order (thanks Codyray1)

v. 1.0 - Initial Release
** Basically works now
** Ultimate minimum of features
** Does regular sprays, animated, fading sprays

v. 0.1 - Initial Commit
** Adding things from my working environment on EC2
** Basic functionality is here, but has several (SEVERAL) bugs


VTFheader.h
#include <string.h>


#ifndef VTFHEADER_H
#define VTFHEADER_H
typedef struct tagVTFHEADER
{
	char		signature[4];		// File signature ("VTF\0").
	unsigned int	version[2];		// version[0].version[1] (currently 7.2).
	unsigned int	headerSize;		// Size of the header struct (16 byte aligned; currently 80 bytes).
	unsigned short	width;			// Width of the largest mipmap in pixels. Must be a power of 2.
	unsigned short	height;			// Height of the largest mipmap in pixels. Must be a power of 2.
	unsigned int	flags;			// VTF flags.
	unsigned short	frames;			// Number of frames, if animated (1 for no animation).
	unsigned short	firstFrame;		// First frame in animation (0 based).
	unsigned char	padding0[4];		// reflectivity padding (16 byte alignment).
	float		reflectivity[3];	// reflectivity vector.
	unsigned char	padding1[4];		// reflectivity padding (8 byte packing).
	float		bumpmapScale;		// Bumpmap scale.
	unsigned int	highResImageFormat;	// High resolution image format.
	unsigned char	mipmapCount;		// Number of mipmaps.
	unsigned int	lowResImageFormat;	// Low resolution image format (always DXT1).
	unsigned char	lowResImageWidth;	// Low resolution image width.
	unsigned char	lowResImageHeight;	// Low resolution image height.
	unsigned short	depth;			// Depth of the largest mipmap in pixels.
						// Must be a power of 2. Can be 0 or 1 for a 2D texture (v7.2 only).
	tagVTFHEADER(){
		strcpy(signature, "VTF");
	}
} VTFHEADER;

#endif /*VTFHEADER_H*/


VTFmain.cpp
#include "VTFnix.cpp"

int main(int argc, char* argv[]){

	if(argc < 3){
		cerr << "Usage:\n";
		cerr << "./VTFnix -i inputFile.ext [-f farImage.ext] [-o outfile.vtf] [-n | --no-alpha]\n";
		cerr << "./VTFnix -i inputDirectory [-f farDirectory] [-o outfile.vtf] [-n | --no-alpha]\n";
		cerr << "Inputs for fading sprays must be identical in size (and number of frames, if an animation)\n";
		return 1;
	} else {
		char *inputFilename = NULL, *farImage = NULL, *outputFilename = NULL;
		bool anim = false;
		bool fade = false;
		bool alpha = true;

		for(int i=1; i<argc; i++){
			if(strcmp(argv[i], "-i") == 0 && i+1 != argc && anim != true){
				inputFilename = argv[i+1];
			} else if(strcmp(argv[i], "-f") == 0 && i+1 != argc && fade != true){
				farImage = argv[i+1];
				fade = true;
			} else if(strcmp(argv[i], "-o") == 0 && i+1 != argc){
				outputFilename = argv[i+1];
			} else if(strcmp(argv[i], "-n") == 0 || strcmp(argv[i], "--no-alpha") == 0){
				alpha = false;
			}
		}


		if(inputFilename == NULL){
			cerr << "Must specify input filename.\n";
			cerr << "Usage:\n";
			cerr << "./VTFnix -i inputFile.ext [-f farImage.ext] [-o outfile.vtf] [-n | --no-alpha]\n";
			cerr << "./VTFnix -i inputDirectory [-f farDirectory] [-o outfile.vtf] [-n | --no-alpha]\n";
			cerr << "Inputs for fading sprays must be identical in size (and number of frames, if an animation)\n";
			return 2;
		}


		if(outputFilename == NULL){
			cerr << "No output file specified, outputting to OUTPUT.vtf\n";
			outputFilename = (char *)"OUTPUT.vtf"; // Workaround typecast to avoid a deprecation warning
		}


		// Convert everything to strings because why not

		string *filename = new string(inputFilename);
		string *out = new string(outputFilename);
		string *far;
		if(farImage != NULL)
			far = new string(farImage);

		if(fade == true){
			fadingImage(*filename, *far, *out, alpha);
		} else {
			animatedImage(*filename, *out, ALL_MIPMAPS, false, alpha); //inputFilename in this case being a folder name
		}

		return 0;			

	}


}


vtfmain.cpp
#include "vtfcmdnix.cpp"

int main(int argc, char* argv[]){

	if(argc < 3){
		cerr << "Usage: -in inputFile.ext [-anim] [-f farImage.ext] [-out outfile.vtf]\n";
		return -1;
	} else {
		char *inputFilename, *farImage, *outputFilename;
		bool anim = false;
		bool fade = false;

		for(int i=1; i<argc; i++){
			if(argv[i] == "-in" && i+1 != argc){
				inputFilename = argv[i+1];
			} else if(argv[i] == "-f" && i+1 != argc){
				farImage = argv[i+1];
				fade = true;
			} else if(argv[i] == "-out" && i+1 != argc){
				outputFilename = argv[i+1];
			} else if(argv[i] == "-anim"){
				anim = true;
			}
		}


		if(inputFilename == NULL){
			cerr << "Must specify input filename.\n";
			cerr << "Usage: -in inputFile.ext [-anim] [-f farImage.ext] [-out outfile.vtf]\n";
			return -2;
		}


		if(outputFilename == NULL){
//			strcpy(outputFilename, inputFilename);
//			strcat(outputFilename, ".vtf");
		}


		cout << &inputFilename;
		if(anim == true){
			singleAnimation(inputFilename, outputFilename, 0, false);
		} else if(fade == true){
			fadingImage(inputFilename, farImage, outputFilename);
		} else {
			singleImage(inputFilename, outputFilename, 0, false);
		}

		return 0;			

	}


}


vtfcmdnix.cpp
#include <cmath>
#include <fstream>
#include <iostream>
#include <nvtt/nvtt.h>
#include <IL/il.h>
#include <IL/ilu.h>
#include "vtfheader.cpp"
#include "vtfoutput.cpp"

using namespace std;

#ifndef VTF_CMD_CPP
#define VTF_CMD_CPP

bool isPowerOfTwo(int x){
	return (x != 0) && ((x & (x - 1)) == 0);
}

void writeLowResData(int imgSize, char* outputFile){
	
	cout << "Writing low res data";

	int size;
	if(imgSize <= 16){
		size = imgSize;
	} else {
		size = 16; //I know my inputs are going to be square
	}

	nvtt::InputOptions inputOptions;
	inputOptions.setTextureLayout(nvtt::TextureType_2D, size, size);
	inputOptions.setMipmapData(ilGetData(), size, size);
	inputOptions.setMipmapGeneration(false);
	
	nvtt::OutputOptions outputOptions;
	Vtfoutput out;
	out.setOutputFile(outputFile);
	outputOptions.setOutputHandler(&out);
	outputOptions.setOutputHeader(false);

	nvtt::CompressionOptions compressionOptions;
	compressionOptions.setFormat(nvtt::Format_DXT1);
	
	nvtt::Compressor compressor;
	compressor.process(inputOptions, compressionOptions, outputOptions);
}

void writeHighResData(int imgSize, char* outputFile){
	
	cout << "Writing high res data" << imgSize << endl;

	int size = imgSize;

	nvtt::InputOptions inputOptions;
	inputOptions.setTextureLayout(nvtt::TextureType_2D, size, size);
	inputOptions.setMipmapData(ilGetData(), size, size);
	inputOptions.setMipmapGeneration(false);

	nvtt::OutputOptions outputOptions;
	Vtfoutput out;
	out.setOutputFile(outputFile);
	outputOptions.setOutputHandler(&out);
	outputOptions.setOutputHeader(false);

	nvtt::CompressionOptions compressionOptions;
	compressionOptions.setFormat(nvtt::Format_DXT5);
	compressionOptions.setQuality(nvtt::Quality_Fastest);
	
	nvtt::Compressor compressor;
	compressor.process(inputOptions, compressionOptions, outputOptions);
}

void writeHeader(int imgSize, int frames, char* outputFile){

	//This is all leftover code, but let's keep it anyway
	VTFHEADER header;
	header.version[0] = 7;
	header.version[1] = 2;
	header.headerSize = 0x50;
	header.width = imgSize;
	header.height = imgSize;
	header.flags = 0x2200;
	header.frames = frames;
	header.firstFrame = 0;
	header.padding0[0] = 0;
	header.padding0[1] = 0;
	header.padding0[2] = 0;
	header.padding0[3] = 0;
	header.reflectivity[0] = 1; //TODO Figure out how the hell to do this
	header.reflectivity[1] = 1; //TODO Figure out how the hell to do this
	header.reflectivity[2] = 1; //TODO Figure out how the hell to do this
	header.padding1[0] = 0;
	header.padding1[1] = 0;
	header.padding1[2] = 0;
	header.padding1[3] = 0;
	header.bumpmapScale = 1;
	header.highResImageFormat = 0xF; //DXT5
	header.mipmapCount = (char)(log2(imgSize) + 1);
	header.lowResImageFormat = 0xD; //DXT1
	if(imgSize <= 16){
		header.lowResImageWidth = imgSize;
		header.lowResImageHeight = imgSize;
	} else {
		header.lowResImageWidth = 16; //I know my inputs are going to be square
		header.lowResImageHeight = 16; //So I don't need to bother with scaling stuff
	}
	header.depth = 1;
	header.padding3[0] = 0;
	header.padding3[1] = 0;
	header.padding3[2] = 0;
	header.resourceCount = 0;
	
	cout << sizeof(tagVTFHEADER);
	// Here's where shit starts to happen
	cout << "WRITING HEADER";
	ofstream output;
	output.open(outputFile, ios::out | ios::binary | ios::app);
	output.write(reinterpret_cast<char *>(&header.signature), 4*sizeof(char));
	output.write(reinterpret_cast<char *>(&header.version), 2*sizeof(int));
	output.write(reinterpret_cast<char *>(&header.headerSize), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.width), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.height), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.flags), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.frames), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.firstFrame), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.padding0), 4*sizeof(char));
	output.write(reinterpret_cast<char *>(&header.reflectivity), 3*sizeof(float));
	output.write(reinterpret_cast<char *>(&header.padding1), 4*sizeof(char));
	output.write(reinterpret_cast<char *>(&header.bumpmapScale), sizeof(float));
	output.write(reinterpret_cast<char *>(&header.highResImageFormat), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.mipmapCount), sizeof(char));
	output.write(reinterpret_cast<char *>(&header.lowResImageFormat), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.lowResImageWidth), sizeof(char));
	output.write(reinterpret_cast<char *>(&header.lowResImageHeight), sizeof(char));
	output.write(reinterpret_cast<char *>(&header.depth), sizeof(short));
	output.write("000000000000000", 15*sizeof(char));
	//output.write(reinterpret_cast<char *>(&header.padding3), 3*sizeof(char));
	//output.write(reinterpret_cast<char *>(&header.resourceCount), sizeof(int));
	output.close();
	
}

int singleImage(char *filename, char *outputFile, int skipBiggest, bool skipHeaderAndLow){
	ilInit();
	iluInit();

	if(!ilLoadImage(filename)){
		cerr << "Error loading input file\n";
		return -1;
	} else if(!isPowerOfTwo(ilGetInteger(IL_IMAGE_WIDTH)) || !isPowerOfTwo(ilGetInteger(IL_IMAGE_HEIGHT))){
		cerr << "Width and height must be powers of two\n";
		return -2;
	} else if(ilGetInteger(IL_IMAGE_WIDTH) != ilGetInteger(IL_IMAGE_HEIGHT)){
		cerr << "Width and height must be equal\n";
		return -3;
	} else {
		
		if(!ilConvertImage(IL_BGRA, IL_UNSIGNED_BYTE)){
			cerr << "Error converting image to BGRA format.\n";
			return -4;
		} else {
		
			ilHint(IL_MEM_SPEED_HINT, IL_LESS_MEM);	
			iluBuildMipmaps();
			int numMips = (int)log2(ilGetInteger(IL_IMAGE_HEIGHT)) + 1;

			if(!skipHeaderAndLow){
				writeHeader(ilGetInteger(IL_IMAGE_HEIGHT), 1, outputFile);

				/* 16x16 image or biggest below that */
				if(numMips >= 5){
					ilActiveMipmap(numMips - 5);
				} else {
					ilActiveMipmap(0);
				}
				writeLowResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile);
			}
	
			int startingMip = numMips-1;
			int biggestMip = 0;
			if(skipBiggest == 1){
				biggestMip = 1;
			} else if(skipBiggest == 2){
				startingMip = 0;
			}
	
			for(int i=startingMip; i>=biggestMip; i--){
				ilLoadImage(filename);
				iluBuildMipmaps();
				ilActiveMipmap(i);
				writeHighResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile);
			}
		}
	cout << "Done.";
	return 0;
	}
}

int singleAnimation(char *filename, char *outputFile, int skipBiggest, bool skipHeaderAndLow){
	ilInit();
	iluInit();

	if(!ilLoadImage(filename)){
		cerr << "Error loading input file\n";
		return -1;
	} else if(!isPowerOfTwo(ilGetInteger(IL_IMAGE_WIDTH)) || !isPowerOfTwo(ilGetInteger(IL_IMAGE_HEIGHT))){
		cerr << "Width and height must be powers of two\n";
		return -2;
	} else if(ilGetInteger(IL_IMAGE_WIDTH) != ilGetInteger(IL_IMAGE_HEIGHT)){
		cerr << "Width and height must be equal\n";
		return -3;
	} else {
		
		if(!ilConvertImage(IL_BGRA, IL_UNSIGNED_BYTE)){
			cerr << "Error converting image to BGRA format.\n";
			return -4;
		} else {
		
			ilHint(IL_MEM_SPEED_HINT, IL_LESS_MEM);	
			int numFrames = ilGetInteger(IL_NUM_IMAGES);
			if(numFrames == 0)
				return singleImage(filename, outputFile, skipBiggest, skipHeaderAndLow);
					
			iluBuildMipmaps();
			int numMips = (int)log2(ilGetInteger(IL_IMAGE_HEIGHT)) + 1;
			
			if(!skipHeaderAndLow){
				writeHeader(ilGetInteger(IL_IMAGE_HEIGHT), numFrames, outputFile);

				/* 16x16 image or biggest below that */
				if(numMips >= 5){
					ilActiveMipmap(numMips - 5);
				} else {
					ilActiveMipmap(0);
				}
				writeLowResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile);
			}

			int startingMip = numMips;
			int biggestMip = 0;
			if(skipBiggest == 1){ //Will skip the largest mipmap
				biggestMip = 1;
			} else if(skipBiggest == 2){ //Will only do the largest mipmap
				startingMip = 1;
			}
			for(int i=startingMip; i>=biggestMip; i--){
				for(int j=0; j<numFrames; j++){
					ILuint image = iluGenImage();
					ilBindImage(image);	

					ilLoadImage(filename);
					ilActiveImage(j);
					iluScale(ilGetInteger(IL_IMAGE_HEIGHT)*2, ilGetInteger(IL_IMAGE_WIDTH)*2, 1);
					iluBuildMipmaps();
					ilActiveMipmap(i);
					if(i != 0){
						writeHighResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile);
						cout << "mip" << i << " frame" << j << " size" << ilGetInteger(IL_IMAGE_HEIGHT) << endl;
					}

					ilDeleteImage(image);
				}
			}
		}
	cout << "Done.";
	return 0;
	}
}


int fadingImage(char *near, char *far, char *outputFile){
	ilInit();
	iluInit();

	singleAnimation(far, outputFile, 1, false);
	singleAnimation(near, outputFile, 2, true);
}


#endif /*VTF_CMD_CPP*/


vtfheader.cpp
#include <string.h>
typedef struct tagVTFHEADER
{
	char		signature[4];		// File signature ("VTF\0").
	unsigned int	version[2];		// version[0].version[1] (currently 7.2).
	unsigned int	headerSize;		// Size of the header struct (16 byte aligned; currently 80 bytes).
	unsigned short	width;			// Width of the largest mipmap in pixels. Must be a power of 2.
	unsigned short	height;			// Height of the largest mipmap in pixels. Must be a power of 2.
	unsigned int	flags;			// VTF flags.
	unsigned short	frames;			// Number of frames, if animated (1 for no animation).
	unsigned short	firstFrame;		// First frame in animation (0 based).
	unsigned char	padding0[4];		// reflectivity padding (16 byte alignment).
	float		reflectivity[3];	// reflectivity vector.
	unsigned char	padding1[4];		// reflectivity padding (8 byte packing).
	float		bumpmapScale;		// Bumpmap scale.
	unsigned int	highResImageFormat;	// High resolution image format.
	unsigned char	mipmapCount;		// Number of mipmaps.
	unsigned int	lowResImageFormat;	// Low resolution image format (always DXT1).
	unsigned char	lowResImageWidth;	// Low resolution image width.
	unsigned char	lowResImageHeight;	// Low resolution image height.
	unsigned short	depth;			// Depth of the largest mipmap in pixels.
						// Must be a power of 2. Can be 0 or 1 for a 2D texture (v7.2 only).


	unsigned char padding3[3];
	unsigned int resourceCount;	
	tagVTFHEADER(){
		strcpy(signature, "VTF");
		//signature = "VTF"
	}
} VTFHEADER;


VTFnix.cpp
#include <cmath>
#include <fstream>
#include <cstdlib>
#include <iostream>
#include <vector>
#include <errno.h>
#include <dirent.h>
#include <algorithm>
#include <nvtt/nvtt.h>
#include <IL/il.h>
#include <IL/ilu.h>
#include "VTFheader.h"
#include "VTFoutput.cpp"

using namespace std;

#ifndef VTFNIX_CPP
#define VTFNIX_CPP

////////////////////////
/// Global constants ///
////////////////////////

const int ALL_MIPMAPS = 0;
const int SKIP_LARGEST_MIPMAP = 1;
const int ONLY_LARGEST_MIPMAP = 2;


////////////////////////
/// Helper functions ///
////////////////////////

bool isPowerOfTwo(int x){
	return (x != 0) && ((x & (x - 1)) == 0);
}

// Taken from the example at:
// http://www.linuxquestions.org/questions/programming-9/c-list-files-in-directory-379323/
int getdir (string dir, vector<string> &files)
{
    DIR *dp;
    struct dirent *dirp;
    if((dp  = opendir(dir.c_str())) == NULL) {
        return errno;
    }

    while ((dirp = readdir(dp)) != NULL) {
    	string fileName = string(dirp->d_name);
    	if(fileName.compare("..") != 0 && fileName.compare(".") != 0)
    		files.push_back(dir + "/" + fileName);
    }

    sort(files.begin(), files.end());

    closedir(dp);
    return 0;
}

// Returns the size of the image (for writing the animation header)
int imageSize(string filename){
	ilInit();

	ILuint image = iluGenImage();
	ilBindImage(image);

	if(!ilLoadImage(filename.c_str())){
		cerr << "Error loading input file\n";
		return 3;
	}

	return ilGetInteger(IL_IMAGE_WIDTH);

	ilDeleteImage(image);
}

/////////////////
/// Functions ///
/////////////////

void writeLowResData(int imgSize, string outputFile){
	
	//cout << "Writing low res data\n";

	int size;
	if(imgSize <= 16){
		size = imgSize;
	} else {
		size = 16; //I know my inputs are going to be square
	}

	const void* imageData = malloc(ilGetInteger(IL_IMAGE_SIZE_OF_DATA));
	//cout << " data_size: " << ilGetInteger(IL_IMAGE_SIZE_OF_DATA) << endl;
	imageData = 0;
	imageData = ilGetData();

	nvtt::InputOptions inputOptions;
	inputOptions.setTextureLayout(nvtt::TextureType_2D, size, size);
	inputOptions.setMipmapData(imageData, size, size);
	inputOptions.setMipmapGeneration(false);
	
	nvtt::OutputOptions outputOptions;
	Vtfoutput out;
	out.setOutputFile(outputFile);
	outputOptions.setOutputHandler(&out);
	outputOptions.setOutputHeader(false);

	nvtt::CompressionOptions compressionOptions;
	compressionOptions.setFormat(nvtt::Format_DXT1);
	compressionOptions.setQuality(nvtt::Quality_Fastest);
	
	nvtt::Compressor compressor;
	compressor.process(inputOptions, compressionOptions, outputOptions);
}

void writeHighResData(int imgSize, string outputFile, bool alpha){
	
	//cout << "Writing high res data for size " << imgSize;

	int size = imgSize;

	const void* imageData = malloc(ilGetInteger(IL_IMAGE_SIZE_OF_DATA));
	//cout << " data_size: " << ilGetInteger(IL_IMAGE_SIZE_OF_DATA) << endl;
	imageData = 0;
	imageData = ilGetData();

	nvtt::InputOptions inputOptions;
	inputOptions.setTextureLayout(nvtt::TextureType_2D, size, size);
	inputOptions.setMipmapData(imageData, size, size);
	inputOptions.setMipmapGeneration(false);

	nvtt::OutputOptions outputOptions;
	Vtfoutput out;
	out.setOutputFile(outputFile);
	outputOptions.setOutputHandler(&out);
	outputOptions.setOutputHeader(false);

	nvtt::CompressionOptions compressionOptions;
	if(alpha)
		compressionOptions.setFormat(nvtt::Format_DXT5);
	else
		compressionOptions.setFormat(nvtt::Format_DXT1);
	compressionOptions.setQuality(nvtt::Quality_Fastest);
	
	nvtt::Compressor compressor;
	compressor.process(inputOptions, compressionOptions, outputOptions);
}

void writeHeader(int imgSize, int frames, string outputFile, bool alpha){

	// A lot of this information comes from the wonderful VTFLib and VTFCMD
	// See vtfheader.cpp for more info on each of these
	VTFHEADER header;
	header.version[0] = 7;
	header.version[1] = 2;
	header.headerSize = 0x50;
	header.width = imgSize;
	header.height = imgSize;
	header.flags = 0x2200;
	header.frames = frames;
	header.firstFrame = 0;
	header.padding0[0] = 0;
	header.padding0[1] = 0;
	header.padding0[2] = 0;
	header.padding0[3] = 0;
	header.reflectivity[0] = 1; //TODO Figure out how to calculate reflectivity
	header.reflectivity[1] = 1;
	header.reflectivity[2] = 1;
	header.padding1[0] = 0;
	header.padding1[1] = 0;
	header.padding1[2] = 0;
	header.padding1[3] = 0;
	header.bumpmapScale = 1;
	if(alpha)
		header.highResImageFormat = 0xF; //DXT5
	else
		header.highResImageFormat = 0xD; //DXT1
	header.mipmapCount = (char)(log2(imgSize) + 1);
	header.lowResImageFormat = 0xD; //Always DXT1
	if(imgSize <= 16){
		header.lowResImageWidth = imgSize;
		header.lowResImageHeight = imgSize;
	} else {
		header.lowResImageWidth = 16; //I know my inputs are going to be square
		header.lowResImageHeight = 16; //So I don't need to bother with scaling stuff
	}
	header.depth = 1;
	
	//cout << "Header size:" + sizeof(tagVTFHEADER);

	// This is the important stuff
	//cout << "Writing Header\n";
	ofstream output;
	output.open(outputFile.c_str(), ios::out | ios::binary);
	output.write(reinterpret_cast<char *>(&header.signature), 4*sizeof(char));
	output.write(reinterpret_cast<char *>(&header.version), 2*sizeof(int));
	output.write(reinterpret_cast<char *>(&header.headerSize), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.width), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.height), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.flags), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.frames), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.firstFrame), sizeof(short));
	output.write(reinterpret_cast<char *>(&header.padding0), 4*sizeof(char));
	output.write(reinterpret_cast<char *>(&header.reflectivity), 3*sizeof(float));
	output.write(reinterpret_cast<char *>(&header.padding1), 4*sizeof(char));
	output.write(reinterpret_cast<char *>(&header.bumpmapScale), sizeof(float));
	output.write(reinterpret_cast<char *>(&header.highResImageFormat), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.mipmapCount), sizeof(char));
	output.write(reinterpret_cast<char *>(&header.lowResImageFormat), sizeof(int));
	output.write(reinterpret_cast<char *>(&header.lowResImageWidth), sizeof(char));
	output.write(reinterpret_cast<char *>(&header.lowResImageHeight), sizeof(char));
	output.write(reinterpret_cast<char *>(&header.depth), sizeof(short));
	output.write("000000000000000", 15*sizeof(char)); //This may not be right.
	output.close();
	
}

int singleImage(string filename, string outputFile, int mipmapOptions, bool onlyHighResData, bool alpha){
	ilInit();
	iluInit();

	if(!ilLoadImage(filename.c_str())){
		cerr << "Error loading input file\n";
		return 3;
	} else if(!isPowerOfTwo(ilGetInteger(IL_IMAGE_WIDTH)) || !isPowerOfTwo(ilGetInteger(IL_IMAGE_HEIGHT))){
		cerr << "Width and height must be powers of two\n";
		return 4;
	} else if(ilGetInteger(IL_IMAGE_WIDTH) != ilGetInteger(IL_IMAGE_HEIGHT)){
		cerr << "Width and height must be equal\n";
		return 5;
	} else {
		
		if(!ilConvertImage(IL_BGRA, IL_UNSIGNED_BYTE)){
			cerr << "Error converting image to BGRA format.\n";
			return 6;
		} else {
		
			ilHint(IL_MEM_SPEED_HINT, IL_LESS_MEM);	
			iluBuildMipmaps();
			int numMips = (int)log2(ilGetInteger(IL_IMAGE_HEIGHT)) + 1;

			if(!onlyHighResData){
				writeHeader(ilGetInteger(IL_IMAGE_HEIGHT), 1, outputFile, alpha);

				/* 16x16 image or biggest below that */
				if(numMips >= 5){
					ilActiveMipmap(numMips - 5);
				} else {
					ilActiveMipmap(0);
				}
				writeLowResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile);
			} else {
				/* Don't ask me why we need this. I don't know.
				I figured since we're regenerating the mips later, it's not an issue.
				But for some reason, if I don't do this, it freaks out and only gives me
				part of my data */
				if(numMips >= 5){
					ilActiveMipmap(numMips - 5);
				} else {
					ilActiveMipmap(0);
				}
			}

			int startingMip; int biggestMip;
			if(mipmapOptions == ALL_MIPMAPS){
				startingMip = numMips-1;
				biggestMip = 0;
			} else if(mipmapOptions == SKIP_LARGEST_MIPMAP){
				startingMip = numMips-1;
				biggestMip = 1;
			} else if(mipmapOptions == ONLY_LARGEST_MIPMAP){
				startingMip = 0;
				biggestMip = 0;
			}
	
			for(int i=startingMip; i>=biggestMip; i--){
				ILuint img = iluGenImage();
				ilBindImage(img);
				ilLoadImage(filename.c_str());
				ilConvertImage(IL_BGRA, IL_UNSIGNED_BYTE);
				iluBuildMipmaps();
				ilActiveMipmap(i);
				writeHighResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile, alpha);
				ilDeleteImage(img);
			}
		}
	//cout << "Done.";
	return 0;
	}
}

int animatedImage(string folder, string outputFile, int mipmapOptions, bool onlyHighResData, bool alpha) {
	vector<string> files;
	int error;
	if ((error = getdir(folder, files)) != 0) {
		if(error == 20){
			clog << "Unable to open input as directory; assuming single image" << endl;
			return singleImage(folder, outputFile, mipmapOptions, onlyHighResData, alpha);
		} else {
	        cerr << "Error(" << error << ") opening " << folder << ": " << strerror(error) << endl;
			return 7;
		}
	}

	int imageSz = imageSize(files.at(0));
	int numMips = (int) log2(imageSz) + 1; //Base our mipmaps on the first image
	int startingMip; int biggestMip;
	if (mipmapOptions == ALL_MIPMAPS) {
		startingMip = numMips - 1;
		biggestMip = 0;
	} else if (mipmapOptions == SKIP_LARGEST_MIPMAP) {
		startingMip = numMips - 1;
		biggestMip = 1;
	} else if (mipmapOptions == ONLY_LARGEST_MIPMAP) {
		startingMip = 0;
		biggestMip = 0;
	}

	for (int i = startingMip; i >= biggestMip; i--) {

		for (int j = 0; j < files.size(); j++) {

			string filename = files.at(j);

			ilInit();
			iluInit();

			if (!ilLoadImage(filename.c_str())) {
				cerr << "Error loading input file\n";
				return 3;
			} else if (!isPowerOfTwo(ilGetInteger(IL_IMAGE_WIDTH)) || !isPowerOfTwo(ilGetInteger(IL_IMAGE_HEIGHT))) {
				cerr << "Width and height must be powers of two\n";
				return 4;
			} else if (ilGetInteger(IL_IMAGE_WIDTH) != ilGetInteger(IL_IMAGE_HEIGHT)) {
				cerr << "Width and height must be equal\n";
				return 5;
			} else {

				if (!ilConvertImage(IL_BGRA, IL_UNSIGNED_BYTE)) {
					cerr << "Error converting image to BGRA format.\n";
					return 6;
				} else {

					ilHint(IL_MEM_SPEED_HINT, IL_LESS_MEM);
					iluBuildMipmaps();
					if (!onlyHighResData && j==0 && i==startingMip) {
						writeHeader(imageSz, files.size(), outputFile, alpha);

						/* 16x16 image or biggest below that */
						if (numMips >= 5) {
							ilActiveMipmap(numMips - 5);
						} else {
							ilActiveMipmap(0);
						}
						writeLowResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile);
					} else {
						/* Don't ask me why we need this. I don't know.
						 I figured since we're regenerating the mips later, it's not an issue.
						 But for some reason, if I don't do this, it freaks out and only gives me
						 part of my data */
						//TODO see if this fixed itself after that last commit
						if (numMips >= 5) {
							ilActiveMipmap(numMips - 5);
						} else {
							ilActiveMipmap(0);
						}
					}

					ILuint img = iluGenImage();
					ilBindImage(img);
					ilLoadImage(filename.c_str());
					ilConvertImage(IL_BGRA, IL_UNSIGNED_BYTE);
					iluBuildMipmaps();
					ilActiveMipmap(i);
					writeHighResData(ilGetInteger(IL_IMAGE_HEIGHT), outputFile, alpha);
					ilDeleteImage(img);
				}
			}
		}

	}

	//cout << "Done.\n";
	return 0;
}


void fadingImage(string near, string far, string outputFile, bool alpha){
	ilInit();
	iluInit();

	animatedImage(far, outputFile, SKIP_LARGEST_MIPMAP, false, alpha); //Output the smallest mipmaps (the far images)
	animatedImage(near, outputFile, ONLY_LARGEST_MIPMAP, true, alpha); //Output the close image (the large mipmap
}


#endif /*VTFNIX_CPP*/
